pkg/asset/kubeconfig/kubeconfig.go

```diff
@@ -0,0 +1,157 @@
+	rootCA        asset.Asset
+	certKey       asset.Asset
+	installConfig asset.Asset
+}
+
+var _ asset.Asset = (*Kubeconfig)(nil)
+
+// Dependencies returns the dependency of the kubeconfig.
+func (k *Kubeconfig) Dependencies() []asset.Asset {
+	return []asset.Asset{
+		k.rootCA,
+		k.certKey,
+		k.installConfig,
+	}
+}
+
+// Generate generates the kubeconfig.
+func (k *Kubeconfig) Generate(parents map[asset.Asset]*asset.State) (*asset.State, error) {
+	var err error
+
+	caCertData, err := getCertKeyData(k.rootCA, parents, ".crt")
+	if err != nil {
+		return nil, err
+	}
+	clientKeyData, err := getCertKeyData(k.certKey, parents, ".key")
+	if err != nil {
+		return nil, err
+	}
+	clientCertData, err := getCertKeyData(k.certKey, parents, ".crt")
+	if err != nil {
+		return nil, err
+	}
+	installConfig, err := getInstallConfig(k.installConfig, parents)
+	if err != nil {
+		return nil, err
+	}
+
+	kubeconfig := clientcmd.Config{
+		Clusters: []clientcmd.NamedCluster{
+			{
+				Name: installConfig.Name,
+				Cluster: clientcmd.Cluster{
+					Server: fmt.Sprintf("https://%s-api.%s:6443", installConfig.Name, installConfig.BaseDomain),
+					CertificateAuthorityData: caCertData,
+				},
+			},
+		},
+		AuthInfos: []clientcmd.NamedAuthInfo{
+			{
+				Name: k.userName,
+				AuthInfo: clientcmd.AuthInfo{
+					ClientCertificateData: clientCertData,
+					ClientKeyData:         clientKeyData,
+				},
+			},
+		},
+		Contexts: []clientcmd.NamedContext{
+			{
+				Name: k.userName,
+				Context: clientcmd.Context{
+					Cluster:  installConfig.Name,
+					AuthInfo: k.userName,
+				},
+			},
+		},
+		CurrentContext: k.userName,
+	}
+
+	data, err := yaml.Marshal(kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	st := &asset.State{
+		Contents: []asset.Content{
+			{
+				Name: assetFilePath(k.rootDir, k.userName),
+				Data: data,
+			},
+		},
+	}
+
+	if err := st.PersistToFile(); err != nil {
+		return nil, err
+	}
+
+	return st, nil
+}
+
+func assetFilePath(rootDir, userName string) string {
+	if userName == KubeconfigUserNamekubelet {
+		return filepath.Join(rootDir, "auth", "kubeconfig-kubelet")
+	}
+	return filepath.Join(rootDir, "auth", "kubeconfig")
+}
+
+// getCertKeyData extracts the cert or key data from the parent map based on the ext (".crt" or ".key").
+// It returns a base64 encoded []bye for the data.
+func getCertKeyData(a asset.Asset, parents map[asset.Asset]*asset.State, ext string) ([]byte, error) {
+	st, ok := parents[a]
+	if !ok {
+		return nil, fmt.Errorf("failed to find %T in parents", a)
+	}
+
+	var data []byte
+	for _, c := range st.Contents {
+		if filepath.Ext(c.Name) == ext {
+			data = c.Data
+			break
+		}
+	}
+	if data == nil {
+		return nil, fmt.Errorf("failed to find data in %v with extension == %q", st, ext)
+	}
+
+	return []byte(base64.StdEncoding.EncodeToString(data)), nil
+}
+
+func getInstallConfig(a asset.Asset, parents map[asset.Asset]*asset.State) (*types.InstallConfig, error) {
```

Comment:
This seems like it should be a public helper defined in `pkg/asset/installconfig`.

Reply:
done